plutok-flat/src/main/kotlin/io/newm/plutok/uplc/flat/FlatEncoder.kt

```diff
@@ -0,0 +1,226 @@
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+import java.math.BigInteger
+import java.nio.ByteBuffer
+import kotlin.math.pow
+
+class FlatEncoder : AbstractEncoder() {
+    override val serializersModule: SerializersModule = EmptySerializersModule()
+
+    private val buffer = ByteBuffer.allocate(1024 * 1000) // 1MB
+    private var usedBits = 0
+    private var currentByte = 0.toUByte()
+
+    override fun encodeValue(value: Any) {
+        when (value) {
+            is Int -> encodeBigInteger(value.toBigInteger())
+            is Long -> encodeBigInteger(value.toBigInteger())
+            is BigInteger -> encodeBigInteger(value)
+            is UByte -> encodeUByte(value)
+            else -> super.encodeValue(value)
+        }
+    }
+
+    /**
+     * Encode 1 unsigned byte.
+     * Uses the next 8 bits in the buffer, can be byte aligned or byte unaligned
+     */
+    fun encodeUByte(value: UByte) {
+        if (usedBits == 0) {
+            currentByte = value
+            nextWord()
+        } else {
+            encodeUByteUnaligned(value)
+        }
+    }
+
+    /**
+     * Encode an arbitrary length integer using zigzag encoding.
+     */
+    fun encodeBigInteger(value: BigInteger) {
+        val i = value.zigZagEncode()
+        encodeWord(i)
+    }
+
+    /**
+     * Write out byte regardless of current buffer alignment.
+     * Write most significant bits in remaining unused bits for the current byte,
+     * then write out the remaining bits at the beginning of the next byte.
+     */
+    private fun encodeUByteUnaligned(value: UByte) {
+        val xShift = (currentByte.toUInt() or (value.toUInt() shr usedBits)).toUByte()
+        buffer.put(xShift.toByte())
+
+        currentByte = (value.toUInt() shl (8 - usedBits)).toUByte()
+    }
+
+    fun encodeListWith(list: List<Any>, encoderFunc: (Any) -> Unit) {
+        for (item in list) {
+            encodeOne()
+            encoderFunc(item)
+        }
+        encodeZero()
+    }
+
+    /**
+     * Encode a unsigned integer of any size.
+     * This is byte alignment agnostic.
+     * We encode the 7 least significant bits of the unsigned byte. If the char value is greater than
+     * 127 we encode a leading 1 followed by repeating the above for the next 7 bits and so on.
+     */
+    fun encodeWord(value: BigInteger) {
+        var d = value
+        do {
+            var w = (d and 127.toBigInteger()).toInt()
+            d = d shr 7
+
+            if (d != BigInteger.ZERO) {
+                w = w or 128
+            }
+            encodeBits(8, w)
+        } while (d != BigInteger.ZERO)
+    }
+
+    /**
+     * Encode a byte in numBits bits with overflow protection.
+     */
+    fun safeEncodeBits(numBits: Int, byte: Int) {
+        if (2.0.pow(numBits.toDouble()) < byte.toDouble()) {
```

Comment:
Maybe a more efficient way here is:
```
if ((1 shl numBits) < byte)
```

Reply:
Added tests and settled on

```
if ((1 shl numBits) - 1 < byte)
```

We need the minus one in there.

Reply:
@rvcas I believe this is a bug in the rust code. To fix it on the rust side, I believe you could just switch to `<=`

```
fn safe_encode_bits(num_bits: u32, byte: u8, e: &mut Encoder) -> Result<(), en::Error> {
    if 2_u8.pow(num_bits) < byte {
        Err(en::Error::Message(format!(
            "Overflow detected, cannot fit {byte} in {num_bits} bits."
        )))
    } else {
        e.bits(num_bits as i64, byte);
        Ok(())
    }
}
```

With this code as-is, if you passed `num_bits=3` and `byte=8`, the number 8 requires 4 bits and should throw an overflow message. But because `2_u8.pow(3) < 8 == false`, it falls through and fails later inside of `e.bits()`